Userland/Libraries/LibJS/Runtime/AsyncFunctionDriverWrapper.cpp

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2021, Ali Mohammad Pur <mpfard@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibJS/Runtime/AsyncFunctionDriverWrapper.h>
+#include <LibJS/Runtime/GlobalObject.h>
+#include <LibJS/Runtime/NativeFunction.h>
+#include <LibJS/Runtime/PromiseReaction.h>
+#include <LibJS/Runtime/VM.h>
+
+namespace JS {
+
+ThrowCompletionOr<Value> AsyncFunctionDriverWrapper::create(GlobalObject& global_object, GeneratorObject* generator_object)
+{
+    auto wrapper = global_object.heap().allocate<AsyncFunctionDriverWrapper>(global_object, global_object, generator_object);
+    return wrapper->react_to_async_task_completion(global_object.vm(), global_object, js_undefined(), true);
+}
+
+AsyncFunctionDriverWrapper::AsyncFunctionDriverWrapper(GlobalObject& global_object, GeneratorObject* generator_object)
+    : Promise(global_object)
+    , m_generator_object(generator_object)
+    , m_on_fulfillment(NativeFunction::create(global_object, "async.on_fulfillment"sv, [this](VM& vm, GlobalObject& global_object) {
+        return react_to_async_task_completion(vm, global_object, vm.argument(0), true);
+    }))
+    , m_on_rejection(NativeFunction::create(global_object, "async.on_rejection"sv, [this](VM& vm, GlobalObject& global_object) {
+        return react_to_async_task_completion(vm, global_object, vm.argument(0), false);
+    }))
+{
+}
+
+ThrowCompletionOr<Value> AsyncFunctionDriverWrapper::react_to_async_task_completion(VM& vm, GlobalObject& global_object, Value value, bool is_successful)
+{
+    auto generator_result = is_successful
+        ? m_generator_object->next_impl(vm, global_object, value, {})
+        : m_generator_object->next_impl(vm, global_object, {}, value);
+
+    if (generator_result.is_throw_completion()) {
+        VERIFY(generator_result.throw_completion().type() == Completion::Type::Throw);
+        vm.clear_exception();
+        auto promise = Promise::create(global_object);
+        promise->reject(generator_result.throw_completion().value());
+        return promise;
+    }
+
+    auto result = generator_result.release_value();
+    if (!result.is_object()) {
+        dbgln("Bad return from next(): {}", result.to_string_without_side_effects());
+        return js_undefined();
+    }
```

Comment:
I don't understand this. If it can happen and we need to handle it, there's no need to log. If it shouldn't happen, just VERIFY().

Random output to stdout breaks the test262 runner btw (which is not using BC at the moment, I know)

Reply:
I forgot to switch that to a VERIFY_NOT_REACHED.
Will fix that